{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Select best dammit annotations for a trinity transcriptome. Create tx2gene maps for use with tximport; aggregate contigs with identical \"best\" annotations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "# requires dammit env (mine-->source activate dammit), or dammit installed in main env\n",
    "# if using an env: if you haven't done so yet, first install an ipy kernel in dammit env:\n",
    "# ipython kernel install --user --name dammit\n",
    "# then start new notebook using \"dammit\" kernel, or for existing nb, select \"switch kernel\" in the Kernel menu\n",
    "# voila! :D\n",
    "from dammit.fileio.gff3 import GFF3Parser"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Link Dammit and Trinity names\n",
    "Use Dammit Namemap to create Dammit_Name : Trinity Transcript/Gene map"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "namemap = \"Trinity.fasta.dammit.namemap.csv\"\n",
    "trin2dammit = pd.read_csv(namemap)\n",
    "\n",
    "# create trinity gene: trans map while we're at it\n",
    "trin2dammit['Trinity_transcript'] = trin2dammit['original'].str.split(' ', 1).str[0]\n",
    "trin2dammit['Trinity_gene'] = trin2dammit['Trinity_transcript'].str.rsplit('_i', 1).str[0] \n",
    "\n",
    "# drop 'original' (long) trinity name; rename dammit name column\n",
    "trin2dammit.rename(index=str, columns={\"renamed\": \"Dammit_transcript\"}, inplace=True)\n",
    "trin2dammit.drop(columns=['original'], inplace=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Parse Dammit GFF3\n",
    "Use Dammit GFF3 to read in annotations, and link with Trinity Names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/tessa/anaconda3/envs/dammit/lib/python3.6/site-packages/dammit/fileio/gff3.py:73: ParserWarning: Both a converter and dtype were specified for column attributes - only the converter will be used\n",
      "  dtype=dict(self.columns)):\n",
      "/Users/tessa/anaconda3/envs/dammit/lib/python3.6/site-packages/dammit/fileio/gff3.py:73: ParserWarning: Both a converter and dtype were specified for column attributes - only the converter will be used\n",
      "  dtype=dict(self.columns)):\n"
     ]
    }
   ],
   "source": [
    "annots_gff3_file = \"Trinity.fasta.dammit.gff3\"\n",
    "annots = GFF3Parser(filename=annots_gff3_file).read() # read in annotation gff3\n",
    "#print(annots.shape)\n",
    "#annots.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Merge annotation gff3 with dammit-trinity namemap"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "annotsTrin = pd.merge(trin2dammit, annots, how='outer', left_on=\"Dammit_transcript\", right_on=\"seqid\")\n",
    "#print(annotsTrin.shape)\n",
    "#annotsTrin.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Of 150663 total transcripts:\n",
      "    63270 transcripts have at least one annotation\n",
      "    87393 transcripts have no annotations\n"
     ]
    }
   ],
   "source": [
    "#count transcripts/annotations & do a sanity check on the merge:\n",
    "numTranscripts = trin2dammit['Dammit_transcript'].nunique()\n",
    "numWithAnnots = annots['seqid'].nunique()\n",
    "numNoAnnots = numTranscripts - numWithAnnots\n",
    "numMergedTranscripts = annotsTrin['Dammit_transcript'].nunique()\n",
    "\n",
    "if (numMergedTranscripts != numTranscripts):\n",
    "    print('something went wrong during merge')\n",
    "else:\n",
    "    print('Of ' + str(numTranscripts) + ' total transcripts:')\n",
    "    print('    ' + str(numWithAnnots) + ' transcripts have at least one annotation')\n",
    "    print('    ' + str(numNoAnnots)  + ' transcripts have no annotations')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Select annotations with best e-values\n",
    "\n",
    "Create annotation csv with the 1. best e-val hit, and 2. best Obimac hit for each transcript"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Dammit_transcript</th>\n",
       "      <th>Trinity_transcript</th>\n",
       "      <th>Trinity_gene</th>\n",
       "      <th>Dbxref</th>\n",
       "      <th>ID</th>\n",
       "      <th>Name</th>\n",
       "      <th>Note</th>\n",
       "      <th>Parent</th>\n",
       "      <th>Target</th>\n",
       "      <th>accuracy</th>\n",
       "      <th>...</th>\n",
       "      <th>end</th>\n",
       "      <th>env_coords</th>\n",
       "      <th>phase</th>\n",
       "      <th>score</th>\n",
       "      <th>seqid</th>\n",
       "      <th>source</th>\n",
       "      <th>start</th>\n",
       "      <th>strand</th>\n",
       "      <th>trunc</th>\n",
       "      <th>type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Transcript_0</td>\n",
       "      <td>TRINITY_DN49110_c0_g1_i1</td>\n",
       "      <td>TRINITY_DN49110_c0_g1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Transcript_1</td>\n",
       "      <td>TRINITY_DN49156_c0_g1_i1</td>\n",
       "      <td>TRINITY_DN49156_c0_g1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Transcript_2</td>\n",
       "      <td>TRINITY_DN49121_c0_g1_i1</td>\n",
       "      <td>TRINITY_DN49121_c0_g1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Transcript_3</td>\n",
       "      <td>TRINITY_DN49116_c0_g1_i1</td>\n",
       "      <td>TRINITY_DN49116_c0_g1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>cds.Transcript_3.p2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Transcript_3.p2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>482.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Transcript_3</td>\n",
       "      <td>transdecoder</td>\n",
       "      <td>173.0</td>\n",
       "      <td>+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CDS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Transcript_3</td>\n",
       "      <td>TRINITY_DN49116_c0_g1_i1</td>\n",
       "      <td>TRINITY_DN49116_c0_g1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Transcript_3.p2.exon1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Transcript_3.p2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>482.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Transcript_3</td>\n",
       "      <td>transdecoder</td>\n",
       "      <td>0.0</td>\n",
       "      <td>+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>exon</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  Dammit_transcript        Trinity_transcript           Trinity_gene Dbxref  \\\n",
       "0      Transcript_0  TRINITY_DN49110_c0_g1_i1  TRINITY_DN49110_c0_g1    NaN   \n",
       "1      Transcript_1  TRINITY_DN49156_c0_g1_i1  TRINITY_DN49156_c0_g1    NaN   \n",
       "2      Transcript_2  TRINITY_DN49121_c0_g1_i1  TRINITY_DN49121_c0_g1    NaN   \n",
       "3      Transcript_3  TRINITY_DN49116_c0_g1_i1  TRINITY_DN49116_c0_g1    NaN   \n",
       "4      Transcript_3  TRINITY_DN49116_c0_g1_i1  TRINITY_DN49116_c0_g1    NaN   \n",
       "\n",
       "                      ID Name Note           Parent Target accuracy  ...   \\\n",
       "0                    NaN  NaN  NaN              NaN    NaN      NaN  ...    \n",
       "1                    NaN  NaN  NaN              NaN    NaN      NaN  ...    \n",
       "2                    NaN  NaN  NaN              NaN    NaN      NaN  ...    \n",
       "3    cds.Transcript_3.p2  NaN  NaN  Transcript_3.p2    NaN      NaN  ...    \n",
       "4  Transcript_3.p2.exon1  NaN  NaN  Transcript_3.p2    NaN      NaN  ...    \n",
       "\n",
       "     end env_coords  phase score         seqid        source  start strand  \\\n",
       "0    NaN        NaN    NaN   NaN           NaN           NaN    NaN    NaN   \n",
       "1    NaN        NaN    NaN   NaN           NaN           NaN    NaN    NaN   \n",
       "2    NaN        NaN    NaN   NaN           NaN           NaN    NaN    NaN   \n",
       "3  482.0        NaN    0.0   NaN  Transcript_3  transdecoder  173.0      +   \n",
       "4  482.0        NaN    NaN   NaN  Transcript_3  transdecoder    0.0      +   \n",
       "\n",
       "   trunc  type  \n",
       "0    NaN   NaN  \n",
       "1    NaN   NaN  \n",
       "2    NaN   NaN  \n",
       "3    NaN   CDS  \n",
       "4    NaN  exon  \n",
       "\n",
       "[5 rows x 22 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "annotsTrin.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(44964, 22)\n",
      "(46472, 26)\n",
      "(150663, 26)\n"
     ]
    }
   ],
   "source": [
    "# find best Obimac hit per transcript (using best evalue)\n",
    "Obimac = annotsTrin[annotsTrin['database'] == \"Obimac_refseq_protein.fa.gz\"]\n",
    "Obimac = Obimac.sort_values(by=['Trinity_transcript', 'score'], ascending=True).query('score < 1e-05').drop_duplicates(subset='Trinity_transcript')[['Trinity_transcript','Name', 'score', 'start', 'end']]\n",
    "#Obimac = Obimac.dropna(axis=0,how=\"all\")\n",
    "Obimac.rename(index=str, columns={'Name': 'Obimac_Name', 'score': 'Obimac_score', 'start': 'Obimac_start', 'end':'Obimac_end'}, inplace=True)\n",
    "\n",
    "#Obimac.tail()\n",
    "\n",
    "#print(Obimac.shape)\n",
    "#Obimac.head()\n",
    "#list(Obimac['Name']) # if you want to read the names\n",
    "# find best Eval that is *not* an Obimac hit\n",
    "\n",
    "noObimac = annotsTrin[annotsTrin['database'] != 'Obimac_refseq_protein.fa.gz']\n",
    "bestEval = noObimac.sort_values(by=['Trinity_transcript', 'score'], ascending=True).query('score < 1e-05').drop_duplicates(subset='Trinity_transcript')\n",
    "#bestEval = bestEval.dropna(axis=0,how=\"all\")\n",
    "\n",
    "print(bestEval.shape)\n",
    "#bestEval.head()\n",
    "# merge bestEval and Obimac annotation info\n",
    "bestAnnots = pd.merge(bestEval, Obimac, how='outer', left_on=\"Trinity_transcript\", right_on=\"Trinity_transcript\")\n",
    "print(bestAnnots.shape)\n",
    "bestAnnots.to_csv(\"bestEvalpertranscript.csv\") # csv of best eval, with Obimac hit if no best eval\n",
    "\n",
    "# to keep the unannotated transcripts, merge those back in\n",
    "# merge will have not know which trinity gene or dammit transcript name to keep: NaN from Annots, or real gene from trin2dammit\n",
    "bestAnnots.drop(columns=['Trinity_gene', \"Dammit_transcript\"], inplace=True) # remove them first, then merge in trin2dammit\n",
    "bestAnnotFull = pd.merge(trin2dammit, bestAnnots, how='outer', on=[\"Trinity_transcript\"])\n",
    "print(bestAnnotFull.shape)\n",
    "bestAnnotFull.to_csv(\"dammit_bestEvalperTranscript.csv\", index=False, sep='\\t')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Dammit_transcript</th>\n",
       "      <th>Trinity_transcript</th>\n",
       "      <th>Trinity_gene</th>\n",
       "      <th>Dbxref</th>\n",
       "      <th>ID</th>\n",
       "      <th>Name</th>\n",
       "      <th>Note</th>\n",
       "      <th>Parent</th>\n",
       "      <th>Target</th>\n",
       "      <th>accuracy</th>\n",
       "      <th>...</th>\n",
       "      <th>seqid</th>\n",
       "      <th>source</th>\n",
       "      <th>start</th>\n",
       "      <th>strand</th>\n",
       "      <th>trunc</th>\n",
       "      <th>type</th>\n",
       "      <th>Obimac_Name</th>\n",
       "      <th>Obimac_score</th>\n",
       "      <th>Obimac_start</th>\n",
       "      <th>Obimac_end</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>150658</th>\n",
       "      <td>Transcript_150658</td>\n",
       "      <td>TRINITY_DN46499_c0_g1_i1</td>\n",
       "      <td>TRINITY_DN46499_c0_g1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>150659</th>\n",
       "      <td>Transcript_150659</td>\n",
       "      <td>TRINITY_DN46470_c0_g1_i1</td>\n",
       "      <td>TRINITY_DN46470_c0_g1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>150660</th>\n",
       "      <td>Transcript_150660</td>\n",
       "      <td>TRINITY_DN46465_c0_g1_i1</td>\n",
       "      <td>TRINITY_DN46465_c0_g1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>homology:113752</td>\n",
       "      <td>RL17_YARLI</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>RL17_YARLI 12 84 +</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>Transcript_150660</td>\n",
       "      <td>LAST</td>\n",
       "      <td>31.0</td>\n",
       "      <td>+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>translated_nucleotide_match</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>150661</th>\n",
       "      <td>Transcript_150661</td>\n",
       "      <td>TRINITY_DN46482_c0_g1_i1</td>\n",
       "      <td>TRINITY_DN46482_c0_g1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>gi|961088763|ref|XP_014771226.1| PREDICTED: so...</td>\n",
       "      <td>1.900000e-13</td>\n",
       "      <td>1.0</td>\n",
       "      <td>33.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>150662</th>\n",
       "      <td>Transcript_150662</td>\n",
       "      <td>TRINITY_DN46441_c0_g1_i1</td>\n",
       "      <td>TRINITY_DN46441_c0_g1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 26 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Dammit_transcript        Trinity_transcript           Trinity_gene  \\\n",
       "150658  Transcript_150658  TRINITY_DN46499_c0_g1_i1  TRINITY_DN46499_c0_g1   \n",
       "150659  Transcript_150659  TRINITY_DN46470_c0_g1_i1  TRINITY_DN46470_c0_g1   \n",
       "150660  Transcript_150660  TRINITY_DN46465_c0_g1_i1  TRINITY_DN46465_c0_g1   \n",
       "150661  Transcript_150661  TRINITY_DN46482_c0_g1_i1  TRINITY_DN46482_c0_g1   \n",
       "150662  Transcript_150662  TRINITY_DN46441_c0_g1_i1  TRINITY_DN46441_c0_g1   \n",
       "\n",
       "       Dbxref               ID        Name Note Parent              Target  \\\n",
       "150658    NaN              NaN         NaN  NaN    NaN                 NaN   \n",
       "150659    NaN              NaN         NaN  NaN    NaN                 NaN   \n",
       "150660    NaN  homology:113752  RL17_YARLI  NaN    NaN  RL17_YARLI 12 84 +   \n",
       "150661    NaN              NaN         NaN  NaN    NaN                 NaN   \n",
       "150662    NaN              NaN         NaN  NaN    NaN                 NaN   \n",
       "\n",
       "       accuracy    ...                  seqid source  start strand  trunc  \\\n",
       "150658      NaN    ...                    NaN    NaN    NaN    NaN    NaN   \n",
       "150659      NaN    ...                    NaN    NaN    NaN    NaN    NaN   \n",
       "150660      NaN    ...      Transcript_150660   LAST   31.0      +    NaN   \n",
       "150661      NaN    ...                    NaN    NaN    NaN    NaN    NaN   \n",
       "150662      NaN    ...                    NaN    NaN    NaN    NaN    NaN   \n",
       "\n",
       "                               type  \\\n",
       "150658                          NaN   \n",
       "150659                          NaN   \n",
       "150660  translated_nucleotide_match   \n",
       "150661                          NaN   \n",
       "150662                          NaN   \n",
       "\n",
       "                                              Obimac_Name  Obimac_score  \\\n",
       "150658                                                NaN           NaN   \n",
       "150659                                                NaN           NaN   \n",
       "150660                                                NaN           NaN   \n",
       "150661  gi|961088763|ref|XP_014771226.1| PREDICTED: so...  1.900000e-13   \n",
       "150662                                                NaN           NaN   \n",
       "\n",
       "        Obimac_start Obimac_end  \n",
       "150658           NaN        NaN  \n",
       "150659           NaN        NaN  \n",
       "150660           NaN        NaN  \n",
       "150661           1.0       33.0  \n",
       "150662           NaN        NaN  \n",
       "\n",
       "[5 rows x 26 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bestAnnotFull.tail()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The above keeps all identified trinity transcripts, and just reports the best annotation (determined by eval), as well as the best Obimac annotation (also det. by eval).\n",
    "\n",
    "However, there are undoubtedly some duplicate annotations, as there are v. likely not that many genes in D. opalescens. For differential expression, we're primarily interested in functional changes, and spreading the expression of a gene across different contigs with the exact same best annotation is not desirable. Can we collapse by best annotation?"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Create Tx2Gene files\n",
    "Create Transcript to Gene (tx2gene) files that assign transcripts to their annotated gene names,\n",
    "rather than just the Trinity gene name. If we want to keep unannotated transcripts, we can retain \n",
    "Trinity gene name for unannotated transcripts. These can be used with tximport to aggregate transcript-level\n",
    "quantification to gene-level counts for differential expression analysis."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "bestAnnot_tx2gene = bestAnnots[['Trinity_transcript', 'Name']].dropna()\n",
    "#bestAnnots['Name'].nunique() # just checking that there are duplicate 'Name' column values. There are.\n",
    "bestAnnot_tx2gene.head()\n",
    "bestAnnot_tx2gene.to_csv('bestEval_tx2gene.txt', index=False, sep = '\\t')\n",
    "# do this using the Obimac best annotations instead:\n",
    "bestObimac_tx2gene = bestAnnots[['Trinity_transcript', 'Obimac_Name']].dropna()\n",
    "bestObimac_tx2gene.to_csv('bestObimac_tx2gene.txt', index=False, sep = '\\t') "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let's make a version that includes unannotated transcripts. Use Trinity gene info as gene name"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(150663, 26)\n",
      "# annotations using only best Evalue: 44964\n",
      "# annotations using best Evalue AND best Obimac Evalue: 46472\n"
     ]
    }
   ],
   "source": [
    "#remake df: (easier for testing)\n",
    "#bestAnnotFull = pd.merge(bestAnnots, trin2dammit, how='outer', on=[\"Trinity_transcript\"])\n",
    "print(bestAnnotFull.shape)\n",
    "#bestAnnotFull.head()\n",
    "# first, fill NA in the 'Name' column with Obimac annotations\n",
    "\n",
    "numAnnots = len(list(bestAnnotFull.Name.dropna()))\n",
    "print('# annotations using only best Evalue: ' + str(numAnnots))\n",
    "bestAnnotFull.Name = bestAnnotFull.Name.fillna(value=bestAnnotFull.Obimac_Name)\n",
    "numAnnots_wOb = len(list(bestAnnotFull.Name.dropna()))\n",
    "print('# annotations using best Evalue AND best Obimac Evalue: ' + str(numAnnots_wOb))\n",
    "\n",
    "bestEandObimac_tx2gene = bestAnnotFull[['Trinity_transcript', 'Name']].dropna()\n",
    "bestEandObimac_tx2gene.to_csv('bestEval_Obimac_tx2gene.txt', index=False, sep = '\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(150663, 2)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Dammit_transcript</th>\n",
       "      <th>Trinity_transcript</th>\n",
       "      <th>Trinity_gene</th>\n",
       "      <th>Dbxref</th>\n",
       "      <th>ID</th>\n",
       "      <th>Name</th>\n",
       "      <th>Note</th>\n",
       "      <th>Parent</th>\n",
       "      <th>Target</th>\n",
       "      <th>accuracy</th>\n",
       "      <th>...</th>\n",
       "      <th>seqid</th>\n",
       "      <th>source</th>\n",
       "      <th>start</th>\n",
       "      <th>strand</th>\n",
       "      <th>trunc</th>\n",
       "      <th>type</th>\n",
       "      <th>Obimac_Name</th>\n",
       "      <th>Obimac_score</th>\n",
       "      <th>Obimac_start</th>\n",
       "      <th>Obimac_end</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Transcript_0</td>\n",
       "      <td>TRINITY_DN49110_c0_g1_i1</td>\n",
       "      <td>TRINITY_DN49110_c0_g1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TRINITY_DN49110_c0_g1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Transcript_1</td>\n",
       "      <td>TRINITY_DN49156_c0_g1_i1</td>\n",
       "      <td>TRINITY_DN49156_c0_g1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TRINITY_DN49156_c0_g1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Transcript_2</td>\n",
       "      <td>TRINITY_DN49121_c0_g1_i1</td>\n",
       "      <td>TRINITY_DN49121_c0_g1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TRINITY_DN49121_c0_g1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Transcript_3</td>\n",
       "      <td>TRINITY_DN49116_c0_g1_i1</td>\n",
       "      <td>TRINITY_DN49116_c0_g1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TRINITY_DN49116_c0_g1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Transcript_4</td>\n",
       "      <td>TRINITY_DN49116_c0_g1_i2</td>\n",
       "      <td>TRINITY_DN49116_c0_g1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TRINITY_DN49116_c0_g1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 26 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  Dammit_transcript        Trinity_transcript           Trinity_gene Dbxref  \\\n",
       "0      Transcript_0  TRINITY_DN49110_c0_g1_i1  TRINITY_DN49110_c0_g1    NaN   \n",
       "1      Transcript_1  TRINITY_DN49156_c0_g1_i1  TRINITY_DN49156_c0_g1    NaN   \n",
       "2      Transcript_2  TRINITY_DN49121_c0_g1_i1  TRINITY_DN49121_c0_g1    NaN   \n",
       "3      Transcript_3  TRINITY_DN49116_c0_g1_i1  TRINITY_DN49116_c0_g1    NaN   \n",
       "4      Transcript_4  TRINITY_DN49116_c0_g1_i2  TRINITY_DN49116_c0_g1    NaN   \n",
       "\n",
       "    ID                   Name Note Parent Target accuracy    ...     seqid  \\\n",
       "0  NaN  TRINITY_DN49110_c0_g1  NaN    NaN    NaN      NaN    ...       NaN   \n",
       "1  NaN  TRINITY_DN49156_c0_g1  NaN    NaN    NaN      NaN    ...       NaN   \n",
       "2  NaN  TRINITY_DN49121_c0_g1  NaN    NaN    NaN      NaN    ...       NaN   \n",
       "3  NaN  TRINITY_DN49116_c0_g1  NaN    NaN    NaN      NaN    ...       NaN   \n",
       "4  NaN  TRINITY_DN49116_c0_g1  NaN    NaN    NaN      NaN    ...       NaN   \n",
       "\n",
       "  source  start strand  trunc  type Obimac_Name Obimac_score  Obimac_start  \\\n",
       "0    NaN    NaN    NaN    NaN   NaN         NaN          NaN           NaN   \n",
       "1    NaN    NaN    NaN    NaN   NaN         NaN          NaN           NaN   \n",
       "2    NaN    NaN    NaN    NaN   NaN         NaN          NaN           NaN   \n",
       "3    NaN    NaN    NaN    NaN   NaN         NaN          NaN           NaN   \n",
       "4    NaN    NaN    NaN    NaN   NaN         NaN          NaN           NaN   \n",
       "\n",
       "  Obimac_end  \n",
       "0        NaN  \n",
       "1        NaN  \n",
       "2        NaN  \n",
       "3        NaN  \n",
       "4        NaN  \n",
       "\n",
       "[5 rows x 26 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Then, fill remaining NA values in the 'Name' column with the Trinity Gene name:\n",
    "bestAnnotFull.Name = bestAnnotFull.Name.fillna(value=bestAnnotFull.Trinity_gene)\n",
    "bestAnnotFull_tx2gene = bestAnnotFull[['Trinity_transcript', 'Name']]\n",
    "print(bestAnnotFull_tx2gene.shape)\n",
    "bestAnnotFull_tx2gene.to_csv('bestEvalFull_tx2gene.txt', index=False, sep = '\\t')\n",
    "bestAnnotFull.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Not used: Aggregate by identical best eval hits:\n",
    "# Collapse using best Eval hit:\n",
    "#bestAnnots.groupby('Name',as_index=False)[['Dammit_transcript']].aggregate(lambda x: list(x))\n",
    "#Obimac.groupby('Name',as_index=False)[['seqid']].aggregate(lambda x: list(x))\n",
    "# this groups by Name, but it it necessary for what I want? I just need a transcript \\t gene, where transcript \n",
    "# is the trinity transcript name, and 'gene' is the 'Name' value."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "dammit",
   "language": "python",
   "name": "dammit"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
