{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Select best dammit annotations for a trinity transcriptome. Create tx2gene maps for use with tximport; aggregate contigs with identical \"best\" annotations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "# requires dammit env (mine-->source activate dammit), or dammit installed in main env\n",
    "# if using an env: if you haven't done so yet, first install an ipy kernel in dammit env:\n",
    "# ipython kernel install --user --name dammit\n",
    "# then start new notebook using \"dammit\" kernel, or for existing nb, select \"switch kernel\" in the Kernel menu\n",
    "# voila! :D\n",
    "from dammit.fileio.gff3 import GFF3Parser"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Link Dammit and Trinity names\n",
    "Use Dammit Namemap to create Dammit_Name : Trinity Transcript/Gene map"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "namemap = \"dammit_w_nr/Trinity.fasta.dammit.namemap.csv\"\n",
    "trin2dammit = pd.read_csv(namemap)\n",
    "\n",
    "# create trinity gene: trans map while we're at it\n",
    "trin2dammit['Trinity_transcript'] = trin2dammit['original'].str.split(' ', 1).str[0]\n",
    "trin2dammit['Trinity_gene'] = trin2dammit['Trinity_transcript'].str.rsplit('_i', 1).str[0] \n",
    "\n",
    "# drop 'original' (long) trinity name; rename dammit name column\n",
    "trin2dammit.rename(index=str, columns={\"renamed\": \"Dammit_transcript\"}, inplace=True)\n",
    "trin2dammit.drop(columns=['original'], inplace=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Parse Dammit GFF3\n",
    "Use Dammit GFF3 to read in annotations, and link with Trinity Names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/tessa/anaconda3/envs/dammit/lib/python3.6/site-packages/dammit/fileio/gff3.py:73: ParserWarning: Both a converter and dtype were specified for column attributes - only the converter will be used\n",
      "  dtype=dict(self.columns)):\n",
      "/Users/tessa/anaconda3/envs/dammit/lib/python3.6/site-packages/dammit/fileio/gff3.py:73: ParserWarning: Both a converter and dtype were specified for column attributes - only the converter will be used\n",
      "  dtype=dict(self.columns)):\n"
     ]
    }
   ],
   "source": [
    "annots_gff3_file = \"dammit_w_nr/Trinity.fasta.dammit.gff3\"\n",
    "annots = GFF3Parser(filename=annots_gff3_file).read() # read in annotation gff3\n",
    "#print(annots.shape)\n",
    "#annots.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Merge annotation gff3 with dammit-trinity namemap"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "annotsTrin = pd.merge(trin2dammit, annots, how='outer', left_on=\"Dammit_transcript\", right_on=\"seqid\")\n",
    "#print(annotsTrin.shape)\n",
    "#annotsTrin.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Of 150663 total transcripts:\n",
      "    70012 transcripts have at least one annotation\n",
      "    80651 transcripts have no annotations\n"
     ]
    }
   ],
   "source": [
    "#count transcripts/annotations & do a sanity check on the merge:\n",
    "numTranscripts = trin2dammit['Dammit_transcript'].nunique()\n",
    "numWithAnnots = annots['seqid'].nunique()\n",
    "numNoAnnots = numTranscripts - numWithAnnots\n",
    "numMergedTranscripts = annotsTrin['Dammit_transcript'].nunique()\n",
    "\n",
    "if (numMergedTranscripts != numTranscripts):\n",
    "    print('something went wrong during merge')\n",
    "else:\n",
    "    print('Of ' + str(numTranscripts) + ' total transcripts:')\n",
    "    print('    ' + str(numWithAnnots) + ' transcripts have at least one annotation')\n",
    "    print('    ' + str(numNoAnnots)  + ' transcripts have no annotations')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Select annotations with best e-values\n",
    "\n",
    "Create annotation csv with the 1. best e-val hit, and 2. best Obimac hit for each transcript"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Dammit_transcript</th>\n",
       "      <th>Trinity_transcript</th>\n",
       "      <th>Trinity_gene</th>\n",
       "      <th>seqid</th>\n",
       "      <th>source</th>\n",
       "      <th>type</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>...</th>\n",
       "      <th>ID</th>\n",
       "      <th>Name</th>\n",
       "      <th>Note</th>\n",
       "      <th>Parent</th>\n",
       "      <th>Target</th>\n",
       "      <th>accuracy</th>\n",
       "      <th>bitscore</th>\n",
       "      <th>database</th>\n",
       "      <th>env_coords</th>\n",
       "      <th>trunc</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Transcript_0</td>\n",
       "      <td>TRINITY_DN49110_c0_g1_i1</td>\n",
       "      <td>TRINITY_DN49110_c0_g1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Transcript_1</td>\n",
       "      <td>TRINITY_DN49156_c0_g1_i1</td>\n",
       "      <td>TRINITY_DN49156_c0_g1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Transcript_2</td>\n",
       "      <td>TRINITY_DN49121_c0_g1_i1</td>\n",
       "      <td>TRINITY_DN49121_c0_g1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Transcript_3</td>\n",
       "      <td>TRINITY_DN49116_c0_g1_i1</td>\n",
       "      <td>TRINITY_DN49116_c0_g1</td>\n",
       "      <td>Transcript_3</td>\n",
       "      <td>transdecoder</td>\n",
       "      <td>CDS</td>\n",
       "      <td>173.0</td>\n",
       "      <td>482.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>...</td>\n",
       "      <td>cds.Transcript_3.p2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Transcript_3.p2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Transcript_3</td>\n",
       "      <td>TRINITY_DN49116_c0_g1_i1</td>\n",
       "      <td>TRINITY_DN49116_c0_g1</td>\n",
       "      <td>Transcript_3</td>\n",
       "      <td>transdecoder</td>\n",
       "      <td>exon</td>\n",
       "      <td>0.0</td>\n",
       "      <td>482.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>...</td>\n",
       "      <td>Transcript_3.p2.exon1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Transcript_3.p2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  Dammit_transcript        Trinity_transcript           Trinity_gene  \\\n",
       "0      Transcript_0  TRINITY_DN49110_c0_g1_i1  TRINITY_DN49110_c0_g1   \n",
       "1      Transcript_1  TRINITY_DN49156_c0_g1_i1  TRINITY_DN49156_c0_g1   \n",
       "2      Transcript_2  TRINITY_DN49121_c0_g1_i1  TRINITY_DN49121_c0_g1   \n",
       "3      Transcript_3  TRINITY_DN49116_c0_g1_i1  TRINITY_DN49116_c0_g1   \n",
       "4      Transcript_3  TRINITY_DN49116_c0_g1_i1  TRINITY_DN49116_c0_g1   \n",
       "\n",
       "          seqid        source  type  start    end  score strand  ...   \\\n",
       "0           NaN           NaN   NaN    NaN    NaN    NaN    NaN  ...    \n",
       "1           NaN           NaN   NaN    NaN    NaN    NaN    NaN  ...    \n",
       "2           NaN           NaN   NaN    NaN    NaN    NaN    NaN  ...    \n",
       "3  Transcript_3  transdecoder   CDS  173.0  482.0    NaN      +  ...    \n",
       "4  Transcript_3  transdecoder  exon    0.0  482.0    NaN      +  ...    \n",
       "\n",
       "                      ID Name Note           Parent Target accuracy bitscore  \\\n",
       "0                    NaN  NaN  NaN              NaN    NaN      NaN      NaN   \n",
       "1                    NaN  NaN  NaN              NaN    NaN      NaN      NaN   \n",
       "2                    NaN  NaN  NaN              NaN    NaN      NaN      NaN   \n",
       "3    cds.Transcript_3.p2  NaN  NaN  Transcript_3.p2    NaN      NaN      NaN   \n",
       "4  Transcript_3.p2.exon1  NaN  NaN  Transcript_3.p2    NaN      NaN      NaN   \n",
       "\n",
       "  database env_coords trunc  \n",
       "0      NaN        NaN   NaN  \n",
       "1      NaN        NaN   NaN  \n",
       "2      NaN        NaN   NaN  \n",
       "3      NaN        NaN   NaN  \n",
       "4      NaN        NaN   NaN  \n",
       "\n",
       "[5 rows x 22 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "annotsTrin.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(54896, 22)\n",
      "(55001, 26)\n",
      "(150663, 26)\n"
     ]
    }
   ],
   "source": [
    "# find best Obimac hit per transcript (using best evalue)\n",
    "Obimac = annotsTrin[annotsTrin['database'] == \"Obimac_refseq_protein.fa.gz\"]\n",
    "Obimac = Obimac.sort_values(by=['Trinity_transcript', 'score'], ascending=True).query('score < 1e-05').drop_duplicates(subset='Trinity_transcript')[['Trinity_transcript','Name', 'score', 'start', 'end']]\n",
    "#Obimac = Obimac.dropna(axis=0,how=\"all\")\n",
    "Obimac.rename(index=str, columns={'Name': 'Obimac_Name', 'score': 'Obimac_score', 'start': 'Obimac_start', 'end':'Obimac_end'}, inplace=True)\n",
    "\n",
    "#Obimac.tail()\n",
    "\n",
    "#print(Obimac.shape)\n",
    "#Obimac.head()\n",
    "#list(Obimac['Name']) # if you want to read the names\n",
    "# find best Eval that is *not* an Obimac hit\n",
    "\n",
    "noObimac = annotsTrin[annotsTrin['database'] != 'Obimac_refseq_protein.fa.gz']\n",
    "bestEval = noObimac.sort_values(by=['Trinity_transcript', 'score'], ascending=True).query('score < 1e-05').drop_duplicates(subset='Trinity_transcript')\n",
    "#bestEval = bestEval.dropna(axis=0,how=\"all\")\n",
    "\n",
    "print(bestEval.shape)\n",
    "#bestEval.head()\n",
    "# merge bestEval and Obimac annotation info\n",
    "bestAnnots = pd.merge(bestEval, Obimac, how='outer', left_on=\"Trinity_transcript\", right_on=\"Trinity_transcript\")\n",
    "print(bestAnnots.shape)\n",
    "bestAnnots.to_csv(\"dammit_w_nr/dammit_bestEvalpertranscript_onlyannots.tsv\", index=False, sep = '\\t') # csv of best eval, with Obimac hit if no best eval\n",
    "\n",
    "# to keep the unannotated transcripts, merge those back in\n",
    "# merge will have not know which trinity gene or dammit transcript name to keep: NaN from Annots, or real gene from trin2dammit\n",
    "bestAnnots.drop(columns=['Trinity_gene', \"Dammit_transcript\"], inplace=True) # remove them first, then merge in trin2dammit\n",
    "bestAnnotFull = pd.merge(trin2dammit, bestAnnots, how='outer', on=[\"Trinity_transcript\"])\n",
    "print(bestAnnotFull.shape)\n",
    "bestAnnotFull.to_csv(\"dammit_w_nr/dammit_bestEvalperTranscript.tsv\", index=False, sep='\\t')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Dammit_transcript</th>\n",
       "      <th>Trinity_transcript</th>\n",
       "      <th>Trinity_gene</th>\n",
       "      <th>seqid</th>\n",
       "      <th>source</th>\n",
       "      <th>type</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>...</th>\n",
       "      <th>Target</th>\n",
       "      <th>accuracy</th>\n",
       "      <th>bitscore</th>\n",
       "      <th>database</th>\n",
       "      <th>env_coords</th>\n",
       "      <th>trunc</th>\n",
       "      <th>Obimac_Name</th>\n",
       "      <th>Obimac_score</th>\n",
       "      <th>Obimac_start</th>\n",
       "      <th>Obimac_end</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>150658</th>\n",
       "      <td>Transcript_150658</td>\n",
       "      <td>TRINITY_DN46499_c0_g1_i1</td>\n",
       "      <td>TRINITY_DN46499_c0_g1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>150659</th>\n",
       "      <td>Transcript_150659</td>\n",
       "      <td>TRINITY_DN46470_c0_g1_i1</td>\n",
       "      <td>TRINITY_DN46470_c0_g1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>150660</th>\n",
       "      <td>Transcript_150660</td>\n",
       "      <td>TRINITY_DN46465_c0_g1_i1</td>\n",
       "      <td>TRINITY_DN46465_c0_g1</td>\n",
       "      <td>Transcript_150660</td>\n",
       "      <td>LAST</td>\n",
       "      <td>translated_nucleotide_match</td>\n",
       "      <td>1.0</td>\n",
       "      <td>247.0</td>\n",
       "      <td>1.500000e-32</td>\n",
       "      <td>+</td>\n",
       "      <td>...</td>\n",
       "      <td>UniRef90_A0CRS7 2 85 +</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>uniref90</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>150661</th>\n",
       "      <td>Transcript_150661</td>\n",
       "      <td>TRINITY_DN46482_c0_g1_i1</td>\n",
       "      <td>TRINITY_DN46482_c0_g1</td>\n",
       "      <td>Transcript_150661</td>\n",
       "      <td>LAST</td>\n",
       "      <td>translated_nucleotide_match</td>\n",
       "      <td>2.0</td>\n",
       "      <td>101.0</td>\n",
       "      <td>2.900000e-17</td>\n",
       "      <td>+</td>\n",
       "      <td>...</td>\n",
       "      <td>UniRef90_Q05973 1489 1522 +</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>uniref90</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>gi|961088763|ref|XP_014771226.1| PREDICTED: so...</td>\n",
       "      <td>1.900000e-13</td>\n",
       "      <td>1.0</td>\n",
       "      <td>33.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>150662</th>\n",
       "      <td>Transcript_150662</td>\n",
       "      <td>TRINITY_DN46441_c0_g1_i1</td>\n",
       "      <td>TRINITY_DN46441_c0_g1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 26 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Dammit_transcript        Trinity_transcript           Trinity_gene  \\\n",
       "150658  Transcript_150658  TRINITY_DN46499_c0_g1_i1  TRINITY_DN46499_c0_g1   \n",
       "150659  Transcript_150659  TRINITY_DN46470_c0_g1_i1  TRINITY_DN46470_c0_g1   \n",
       "150660  Transcript_150660  TRINITY_DN46465_c0_g1_i1  TRINITY_DN46465_c0_g1   \n",
       "150661  Transcript_150661  TRINITY_DN46482_c0_g1_i1  TRINITY_DN46482_c0_g1   \n",
       "150662  Transcript_150662  TRINITY_DN46441_c0_g1_i1  TRINITY_DN46441_c0_g1   \n",
       "\n",
       "                    seqid source                         type  start    end  \\\n",
       "150658                NaN    NaN                          NaN    NaN    NaN   \n",
       "150659                NaN    NaN                          NaN    NaN    NaN   \n",
       "150660  Transcript_150660   LAST  translated_nucleotide_match    1.0  247.0   \n",
       "150661  Transcript_150661   LAST  translated_nucleotide_match    2.0  101.0   \n",
       "150662                NaN    NaN                          NaN    NaN    NaN   \n",
       "\n",
       "               score strand    ...                           Target accuracy  \\\n",
       "150658           NaN    NaN    ...                              NaN      NaN   \n",
       "150659           NaN    NaN    ...                              NaN      NaN   \n",
       "150660  1.500000e-32      +    ...           UniRef90_A0CRS7 2 85 +      NaN   \n",
       "150661  2.900000e-17      +    ...      UniRef90_Q05973 1489 1522 +      NaN   \n",
       "150662           NaN    NaN    ...                              NaN      NaN   \n",
       "\n",
       "       bitscore  database env_coords trunc  \\\n",
       "150658      NaN       NaN        NaN   NaN   \n",
       "150659      NaN       NaN        NaN   NaN   \n",
       "150660      NaN  uniref90        NaN   NaN   \n",
       "150661      NaN  uniref90        NaN   NaN   \n",
       "150662      NaN       NaN        NaN   NaN   \n",
       "\n",
       "                                              Obimac_Name  Obimac_score  \\\n",
       "150658                                                NaN           NaN   \n",
       "150659                                                NaN           NaN   \n",
       "150660                                                NaN           NaN   \n",
       "150661  gi|961088763|ref|XP_014771226.1| PREDICTED: so...  1.900000e-13   \n",
       "150662                                                NaN           NaN   \n",
       "\n",
       "       Obimac_start Obimac_end  \n",
       "150658          NaN        NaN  \n",
       "150659          NaN        NaN  \n",
       "150660          NaN        NaN  \n",
       "150661          1.0       33.0  \n",
       "150662          NaN        NaN  \n",
       "\n",
       "[5 rows x 26 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bestAnnotFull.tail()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The above keeps all identified trinity transcripts, and just reports the best annotation (determined by eval), as well as the best Obimac annotation (also det. by eval).\n",
    "\n",
    "However, there are undoubtedly some duplicate annotations, as there are v. likely not that many genes in D. opalescens. For differential expression, we're primarily interested in functional changes, and spreading the expression of a gene across different contigs with the exact same best annotation is not desirable. Can we collapse by best annotation?"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Create Tx2Gene files\n",
    "Create Transcript to Gene (tx2gene) files that assign transcripts to their annotated gene names,\n",
    "rather than just the Trinity gene name. If we want to keep unannotated transcripts, we can retain \n",
    "Trinity gene name for unannotated transcripts. These can be used with tximport to aggregate transcript-level\n",
    "quantification to gene-level counts for differential expression analysis."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "bestAnnot_tx2gene = bestAnnots[['Trinity_transcript', 'Name']].dropna()\n",
    "#bestAnnots['Name'].nunique() # just checking that there are duplicate 'Name' column values. There are.\n",
    "bestAnnot_tx2gene.head()\n",
    "bestAnnot_tx2gene.to_csv('dammit_w_nr/bestEval_tx2gene.txt', index=False, sep = '\\t')\n",
    "# do this using the Obimac best annotations instead:\n",
    "bestObimac_tx2gene = bestAnnots[['Trinity_transcript', 'Obimac_Name']].dropna()\n",
    "bestObimac_tx2gene.to_csv('dammit_w_nr/bestObimac_tx2gene.txt', index=False, sep = '\\t') "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let's make a version that includes unannotated transcripts. Use Trinity gene info as gene name"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(150663, 26)\n",
      "# annotations using only best Evalue: 54896\n",
      "# annotations using best Evalue AND best Obimac Evalue: 55001\n"
     ]
    }
   ],
   "source": [
    "#remake df: (easier for testing)\n",
    "bestAnnotFull = pd.merge(bestAnnots, trin2dammit, how='outer', on=[\"Trinity_transcript\"])\n",
    "print(bestAnnotFull.shape)\n",
    "#bestAnnotFull.head()\n",
    "# first, fill NA in the 'Name' column with Obimac annotations\n",
    "\n",
    "numAnnots = len(list(bestAnnotFull.Name.dropna()))\n",
    "print('# annotations using only best Evalue: ' + str(numAnnots))\n",
    "bestAnnotFull.Name = bestAnnotFull.Name.fillna(value=bestAnnotFull.Obimac_Name)\n",
    "numAnnots_wOb = len(list(bestAnnotFull.Name.dropna()))\n",
    "print('# annotations using best Evalue AND best Obimac Evalue: ' + str(numAnnots_wOb))\n",
    "\n",
    "bestEandObimac_tx2gene = bestAnnotFull[['Trinity_transcript', 'Name']].dropna()\n",
    "bestEandObimac_tx2gene.to_csv('dammit_w_nr/bestEval_Obimac_tx2gene.txt', index=False, sep = '\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(150663, 2)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Trinity_transcript</th>\n",
       "      <th>seqid</th>\n",
       "      <th>source</th>\n",
       "      <th>type</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>phase</th>\n",
       "      <th>Dbxref</th>\n",
       "      <th>...</th>\n",
       "      <th>bitscore</th>\n",
       "      <th>database</th>\n",
       "      <th>env_coords</th>\n",
       "      <th>trunc</th>\n",
       "      <th>Obimac_Name</th>\n",
       "      <th>Obimac_score</th>\n",
       "      <th>Obimac_start</th>\n",
       "      <th>Obimac_end</th>\n",
       "      <th>Dammit_transcript</th>\n",
       "      <th>Trinity_gene</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TRINITY_DN10002_c0_g1_i1</td>\n",
       "      <td>Transcript_55709</td>\n",
       "      <td>LAST</td>\n",
       "      <td>translated_nucleotide_match</td>\n",
       "      <td>1.0</td>\n",
       "      <td>400.0</td>\n",
       "      <td>3.000000e-109</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>uniref90</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Transcript_55709</td>\n",
       "      <td>TRINITY_DN10002_c0_g1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>TRINITY_DN10005_c0_g1_i1</td>\n",
       "      <td>Transcript_55708</td>\n",
       "      <td>LAST</td>\n",
       "      <td>translated_nucleotide_match</td>\n",
       "      <td>7.0</td>\n",
       "      <td>160.0</td>\n",
       "      <td>2.600000e-14</td>\n",
       "      <td>+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>uniref90</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Transcript_55708</td>\n",
       "      <td>TRINITY_DN10005_c0_g1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>TRINITY_DN10017_c0_g1_i1</td>\n",
       "      <td>Transcript_55702</td>\n",
       "      <td>HMMER</td>\n",
       "      <td>protein_hmm_match</td>\n",
       "      <td>141.0</td>\n",
       "      <td>345.0</td>\n",
       "      <td>9.900000e-16</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>\"Pfam:PF00085.16\"</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>127 345</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Transcript_55702</td>\n",
       "      <td>TRINITY_DN10017_c0_g1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>TRINITY_DN10022_c0_g1_i1</td>\n",
       "      <td>Transcript_55699</td>\n",
       "      <td>LAST</td>\n",
       "      <td>translated_nucleotide_match</td>\n",
       "      <td>97.0</td>\n",
       "      <td>544.0</td>\n",
       "      <td>2.300000e-57</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>uniref90</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Transcript_55699</td>\n",
       "      <td>TRINITY_DN10022_c0_g1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>TRINITY_DN10022_c0_g1_i2</td>\n",
       "      <td>Transcript_55700</td>\n",
       "      <td>LAST</td>\n",
       "      <td>translated_nucleotide_match</td>\n",
       "      <td>97.0</td>\n",
       "      <td>187.0</td>\n",
       "      <td>1.800000e-11</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>uniref90</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>gi|961088520|ref|XP_014771140.1| PREDICTED: po...</td>\n",
       "      <td>1.400000e-11</td>\n",
       "      <td>32.0</td>\n",
       "      <td>62.0</td>\n",
       "      <td>Transcript_55700</td>\n",
       "      <td>TRINITY_DN10022_c0_g1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 26 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         Trinity_transcript             seqid source  \\\n",
       "0  TRINITY_DN10002_c0_g1_i1  Transcript_55709   LAST   \n",
       "1  TRINITY_DN10005_c0_g1_i1  Transcript_55708   LAST   \n",
       "2  TRINITY_DN10017_c0_g1_i1  Transcript_55702  HMMER   \n",
       "3  TRINITY_DN10022_c0_g1_i1  Transcript_55699   LAST   \n",
       "4  TRINITY_DN10022_c0_g1_i2  Transcript_55700   LAST   \n",
       "\n",
       "                          type  start    end          score strand  phase  \\\n",
       "0  translated_nucleotide_match    1.0  400.0  3.000000e-109      -    NaN   \n",
       "1  translated_nucleotide_match    7.0  160.0   2.600000e-14      +    NaN   \n",
       "2            protein_hmm_match  141.0  345.0   9.900000e-16    NaN    NaN   \n",
       "3  translated_nucleotide_match   97.0  544.0   2.300000e-57      -    NaN   \n",
       "4  translated_nucleotide_match   97.0  187.0   1.800000e-11      -    NaN   \n",
       "\n",
       "              Dbxref          ...           bitscore  database env_coords  \\\n",
       "0                NaN          ...                NaN  uniref90        NaN   \n",
       "1                NaN          ...                NaN  uniref90        NaN   \n",
       "2  \"Pfam:PF00085.16\"          ...                NaN       NaN    127 345   \n",
       "3                NaN          ...                NaN  uniref90        NaN   \n",
       "4                NaN          ...                NaN  uniref90        NaN   \n",
       "\n",
       "  trunc                                        Obimac_Name  Obimac_score  \\\n",
       "0   NaN                                                NaN           NaN   \n",
       "1   NaN                                                NaN           NaN   \n",
       "2   NaN                                                NaN           NaN   \n",
       "3   NaN                                                NaN           NaN   \n",
       "4   NaN  gi|961088520|ref|XP_014771140.1| PREDICTED: po...  1.400000e-11   \n",
       "\n",
       "  Obimac_start Obimac_end Dammit_transcript           Trinity_gene  \n",
       "0          NaN        NaN  Transcript_55709  TRINITY_DN10002_c0_g1  \n",
       "1          NaN        NaN  Transcript_55708  TRINITY_DN10005_c0_g1  \n",
       "2          NaN        NaN  Transcript_55702  TRINITY_DN10017_c0_g1  \n",
       "3          NaN        NaN  Transcript_55699  TRINITY_DN10022_c0_g1  \n",
       "4         32.0       62.0  Transcript_55700  TRINITY_DN10022_c0_g1  \n",
       "\n",
       "[5 rows x 26 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Then, fill remaining NA values in the 'Name' column with the Trinity Gene name:\n",
    "bestAnnotFull.Name = bestAnnotFull.Name.fillna(value=bestAnnotFull.Trinity_gene)\n",
    "bestAnnotFull_tx2gene = bestAnnotFull[['Trinity_transcript', 'Name']]\n",
    "print(bestAnnotFull_tx2gene.shape)\n",
    "bestAnnotFull_tx2gene.to_csv('dammit_w_nr/bestEvalFull_tx2gene.txt', index=False, sep = '\\t')\n",
    "bestAnnotFull.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "#---> WORKING HERE NOW\n",
    "# try to collapse annots using best annot database --> I think OrthoDB? Alt we could check if ANY of them are the same. \n",
    "# any of them --> would need to collapse based on \"Name column\" rather than on score... hmmmmm can try that! \n",
    "# assuming all of them are decent e-values, this could work. \n",
    "#WHAT ABOUT BEST EVAL ON ORTHO DB (or any other db, just want to choose a primary one for collapsing purposes)\n",
    "#... really want to collapse if Any annotations are the same..."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "# annotations using only best Ortho: 41250\n",
      "# annotations using only best Ortho + best Eval: 54896\n",
      "# annotations using best OrthoDB, Evalue, best Obimac Evalue: 55001\n",
      "(150663, 30)\n"
     ]
    }
   ],
   "source": [
    "# best OrthoDB hit\n",
    "ortho = annotsTrin[annotsTrin['database'] == 'OrthoDB']\n",
    "bestOrtho = ortho.sort_values(by=['Trinity_transcript', 'score'], ascending=True).query('score < 1e-05').drop_duplicates(subset='Trinity_transcript')\n",
    "numAnnotsOrtho = len(list(bestOrtho.Name.dropna()))\n",
    "print('# annotations using only best Ortho: ' + str(numAnnotsOrtho))\n",
    "\n",
    "# remake bestEval (for better testing)\n",
    "noObimac = annotsTrin[annotsTrin['database'] != 'Obimac_refseq_protein.fa.gz']\n",
    "bestEval = noObimac.sort_values(by=['Trinity_transcript', 'score'], ascending=True).query('score < 1e-05').drop_duplicates(subset='Trinity_transcript')\n",
    "# subset by columns\n",
    "bestEval = bestEval[['Trinity_transcript','Name', 'score', 'start', 'end']]\n",
    "bestEval.rename(index=str, columns={'Name': 'bEval_Name', 'score': 'bEval_score', 'start': 'bEval_start', 'end':'bEval_end'}, inplace=True)\n",
    "#merge\n",
    "bestOrthoEv = pd.merge(bestEval, bestOrtho, how='outer', on=\"Trinity_transcript\")\n",
    "numAnnotsOrthoEv = len(list(bestOrthoEv.Trinity_transcript.dropna()))\n",
    "print('# annotations using only best Ortho + best Eval: ' + str(numAnnotsOrthoEv))\n",
    "bestOrthoEv.head()\n",
    "#merge with Obimac hits\n",
    "bestOrthoEvOb = pd.merge(bestOrthoEv, Obimac, how='outer', on=\"Trinity_transcript\")\n",
    "numAnnots_wEv_Ob = len(list(bestOrthoEvOb.Trinity_transcript.dropna()))\n",
    "print('# annotations using best OrthoDB, Evalue, best Obimac Evalue: ' + str(numAnnots_wEv_Ob))\n",
    "\n",
    "# ok, now add in the unannotated hits\n",
    "bestOrthoEvOb.drop(columns=['Trinity_gene', \"Dammit_transcript\"], inplace=True) # remove them first, then merge in trin2dammit\n",
    "bestOrthoEvFull = pd.merge(trin2dammit, bestOrthoEvOb, how='outer', on=[\"Trinity_transcript\"])\n",
    "print(bestOrthoEvFull.shape)\n",
    "bestOrthoEvFull.to_csv(\"dammit_w_nr/dammit_bestOrtho_Ev_Ob_perTranscript.tsv\", index=False, sep='\\t')\n",
    "\n",
    "\n",
    "# To generate the appropriate tx2gene files:\n",
    "#fill NA's in the 'Name' column with 1. best Eval, 2. Ob, 3, Trinity gene name to Name \n",
    "bestOrthoEvFull.Name = bestOrthoEvFull.Name.fillna(value=bestOrthoEvFull.bEval_Name)\n",
    "bestOrthoEvFull.Name = bestOrthoEvFull.Name.fillna(value=bestOrthoEvFull.Obimac_Name)\n",
    "bestOrthoEvFull.Name = bestOrthoEvFull.Name.fillna(value=bestOrthoEvFull.Trinity_gene)\n",
    "# now just take the Transcript, Name cols:\n",
    "bestOrthoEvFull_tx2gene = bestOrthoEvFull[['Trinity_transcript', 'Name']]\n",
    "bestOrthoEvFull_tx2gene.to_csv('dammit_w_nr/bestOrtho_EvOb_tx2gene.txt', index=False, sep = '\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Trinity_transcript</th>\n",
       "      <th>Name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>150658</th>\n",
       "      <td>TRINITY_DN46499_c0_g1_i1</td>\n",
       "      <td>TRINITY_DN46499_c0_g1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>150659</th>\n",
       "      <td>TRINITY_DN46470_c0_g1_i1</td>\n",
       "      <td>TRINITY_DN46470_c0_g1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>150660</th>\n",
       "      <td>TRINITY_DN46465_c0_g1_i1</td>\n",
       "      <td>RL17_YARLI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>150661</th>\n",
       "      <td>TRINITY_DN46482_c0_g1_i1</td>\n",
       "      <td>UniRef90_Q05973</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>150662</th>\n",
       "      <td>TRINITY_DN46441_c0_g1_i1</td>\n",
       "      <td>TRINITY_DN46441_c0_g1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              Trinity_transcript                   Name\n",
       "150658  TRINITY_DN46499_c0_g1_i1  TRINITY_DN46499_c0_g1\n",
       "150659  TRINITY_DN46470_c0_g1_i1  TRINITY_DN46470_c0_g1\n",
       "150660  TRINITY_DN46465_c0_g1_i1             RL17_YARLI\n",
       "150661  TRINITY_DN46482_c0_g1_i1        UniRef90_Q05973\n",
       "150662  TRINITY_DN46441_c0_g1_i1  TRINITY_DN46441_c0_g1"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bestOrthoEvFull_tx2gene.tail()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "# annotations using only best Obimac: 21254\n",
      "# annotations using only best Obimac + best Eval: 55001\n",
      "(150663, 26)\n"
     ]
    }
   ],
   "source": [
    "# Let's try Ob annotation 1st, then best Eval, then Trinity Gene name:\n",
    "# best Obimac hit\n",
    "Obimac = annotsTrin[annotsTrin['database'] == \"Obimac_refseq_protein.fa.gz\"]\n",
    "Obimac = Obimac.sort_values(by=['Trinity_transcript', 'score'], ascending=True).query('score < 1e-05').drop_duplicates(subset='Trinity_transcript')#[['Trinity_transcript','Name', 'score', 'start', 'end']]\n",
    "numAnnotsOb = len(list(Obimac.Trinity_transcript.dropna()))\n",
    "print('# annotations using only best Obimac: ' + str(numAnnotsOb))\n",
    "#Obimac = Obimac.dropna(axis=0,how=\"all\")\n",
    "# no need to rename, since they're going to be primary hit this time\n",
    "#Obimac.rename(index=str, columns={'Name': 'Obimac_Name', 'score': 'Obimac_score', 'start': 'Obimac_start', 'end':'Obimac_end'}, inplace=True)\n",
    "#Obimac.tail()\n",
    "# remake bestEval (for better testing)\n",
    "noObimac = annotsTrin[annotsTrin['database'] != 'Obimac_refseq_protein.fa.gz']\n",
    "bestEval = noObimac.sort_values(by=['Trinity_transcript', 'score'], ascending=True).query('score < 1e-05').drop_duplicates(subset='Trinity_transcript')\n",
    "# subset by columns\n",
    "bestEval = bestEval[['Trinity_transcript','Name', 'score', 'start', 'end']]\n",
    "bestEval.rename(index=str, columns={'Name': 'bEval_Name', 'score': 'bEval_score', 'start': 'bEval_start', 'end':'bEval_end'}, inplace=True)\n",
    "#merge\n",
    "bestObEv = pd.merge(Obimac, bestEval, how='outer', on=\"Trinity_transcript\")\n",
    "numAnnotsObEv = len(list(bestObEv.Trinity_transcript.dropna()))\n",
    "print('# annotations using only best Obimac + best Eval: ' + str(numAnnotsObEv))\n",
    "bestObEv.head()\n",
    "# ok, now add in the unannotated hits\n",
    "bestObEv.drop(columns=['Trinity_gene', \"Dammit_transcript\"], inplace=True) # remove them first, then merge in trin2dammit\n",
    "bestObEvFull = pd.merge(trin2dammit, bestObEv, how='outer', on=[\"Trinity_transcript\"])\n",
    "print(bestObEvFull.shape)\n",
    "bestObEvFull.to_csv(\"dammit_w_nr/dammit_bestObEvannot.tsv\", index=False, sep='\\t')\n",
    "\n",
    "\n",
    "# To generate the appropriate tx2gene files:\n",
    "#fill NA's in the 'Name' column with 1. best Eval, 2. Ob, 3, Trinity gene name to Name \n",
    "bestObEvFull.Name = bestObEvFull.Name.fillna(value=bestObEvFull.bEval_Name)\n",
    "bestObEvFull.Name = bestObEvFull.Name.fillna(value=bestObEvFull.Trinity_gene)\n",
    "# now just take the Transcript, Name cols:\n",
    "bestObEvFull_tx2gene = bestObEvFull[['Trinity_transcript', 'Name']]\n",
    "bestObEvFull_tx2gene.to_csv('dammit_w_nr/bestObEv_tx2gene.txt', index=False, sep = '\\t')\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "# annotations using only best Obimac: 21254\n",
      "# annotations using only best Obimac + best ur90: 51974\n",
      "# annotations using best Obimac + best ur90 + best Eval: 55001\n",
      "(150663, 30)\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "Index(['Trinity_transcript', 'seqid', 'source', 'type', 'start', 'end',\n",
       "       'score', 'strand', 'phase', 'Dbxref', 'ID', 'Name', 'Note', 'Parent',\n",
       "       'Target', 'accuracy', 'bitscore', 'database', 'env_coords', 'trunc',\n",
       "       'ur90_Name', 'ur90_score', 'ur90_start', 'ur90_end', 'bEval_Name',\n",
       "       'bEval_score', 'bEval_start', 'bEval_end'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# That worked better! Let's try: Ob -> Uniref90 --> bestEval --> Trinity Gene\n",
    "# Let's try Ob annotation 1st, then best Eval, then Trinity Gene name:\n",
    "# best Obimac hit\n",
    "Obimac = annotsTrin[annotsTrin['database'] == \"Obimac_refseq_protein.fa.gz\"]\n",
    "Obimac = Obimac.sort_values(by=['Trinity_transcript', 'score'], ascending=True).query('score < 1e-05').drop_duplicates(subset='Trinity_transcript')#[['Trinity_transcript','Name', 'score', 'start', 'end']]\n",
    "numAnnotsOb = len(list(Obimac.Trinity_transcript.dropna()))\n",
    "print('# annotations using only best Obimac: ' + str(numAnnotsOb))\n",
    "#Uniref90\n",
    "ur90 = annotsTrin[annotsTrin['database'] == \"uniref90\"]\n",
    "ur90 = ur90.sort_values(by=['Trinity_transcript', 'score'], ascending=True).query('score < 1e-05').drop_duplicates(subset='Trinity_transcript')[['Trinity_transcript','Name', 'score', 'start', 'end']]\n",
    "ur90.rename(index=str, columns={'Name': 'ur90_Name', 'score': 'ur90_score', 'start': 'ur90_start', 'end':'ur90_end'}, inplace=True)\n",
    "\n",
    "# remake bestEval (for better testing)\n",
    "noObimac = annotsTrin[annotsTrin['database'] != 'Obimac_refseq_protein.fa.gz']\n",
    "noObUr90 = noObimac[noObimac['database']!= 'uniref90']\n",
    "bestE = noObUr90.sort_values(by=['Trinity_transcript', 'score'], ascending=True).query('score < 1e-05').drop_duplicates(subset='Trinity_transcript')\n",
    "# subset by columns\n",
    "bestE = bestE[['Trinity_transcript','Name', 'score', 'start', 'end']]\n",
    "bestE.rename(index=str, columns={'Name': 'bEval_Name', 'score': 'bEval_score', 'start': 'bEval_start', 'end':'bEval_end'}, inplace=True)\n",
    "\n",
    "\n",
    "#ok, now we have Ob, ur90, and bestE. merge.\n",
    "bestOb90 = pd.merge(Obimac, ur90, how='outer', on=\"Trinity_transcript\")\n",
    "numAnnotsOb90 = len(list(bestOb90.Trinity_transcript.dropna()))\n",
    "print('# annotations using only best Obimac + best ur90: ' + str(numAnnotsOb90))\n",
    "bestOb90e =  pd.merge(bestOb90, bestE, how='outer', on=\"Trinity_transcript\")\n",
    "numAnnotsOb90e = len(list(bestOb90e.Trinity_transcript.dropna()))\n",
    "print('# annotations using best Obimac + best ur90 + best Eval: ' + str(numAnnotsOb90e))\n",
    "\n",
    "bestOb90e.head()\n",
    "#print tx2gene of only Annotated contigs:\n",
    "bestOb90e_tx2gene = bestOb90e[['Trinity_transcript', 'Name']]\n",
    "bestOb90e_tx2gene.to_csv(\"dammit_w_nr/dammit_bestOb_ur90Ev_onlyAnnots_tx2gene.txt\", index=False, sep='\\t')\n",
    "\n",
    "#ok, now add in the unannotated hits\n",
    "bestOb90e.drop(columns=['Trinity_gene', \"Dammit_transcript\"], inplace=True) # remove them first, then merge in trin2dammit\n",
    "bestOb90eFull = pd.merge(trin2dammit, bestOb90e, how='outer', on=[\"Trinity_transcript\"])\n",
    "print(bestOb90eFull.shape)\n",
    "bestOb90eFull.to_csv(\"dammit_w_nr/dammit_bestOb_ur90Evannot.tsv\", index=False, sep='\\t')\n",
    "\n",
    "\n",
    "# To generate the appropriate tx2gene files:\n",
    "#fill NA's in the 'Name' column with 1. best Eval, 2. Ob, 3, Trinity gene name to Name \n",
    "bestOb90eFull.Name = bestOb90eFull.Name.fillna(value=bestOb90eFull.ur90_Name)\n",
    "bestOb90eFull.Name = bestOb90eFull.Name.fillna(value=bestOb90eFull.bEval_Name)\n",
    "bestOb90eFull.Name = bestOb90eFull.Name.fillna(value=bestOb90eFull.Trinity_gene)\n",
    "# now just take the Transcript, Name cols:\n",
    "bestOb90eFull_tx2gene = bestOb90eFull[['Trinity_transcript', 'Name']]\n",
    "bestOb90eFull_tx2gene.to_csv('dammit_w_nr/bestOb_ur90Ev_tx2gene.txt', index=False, sep = '\\t')\n",
    "\n",
    "\n",
    "bestOb90e.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Not used: Aggregate by identical best eval hits:\n",
    "# Collapse using best Eval hit:\n",
    "#bestAnnots.groupby('Name',as_index=False)[['Dammit_transcript']].aggregate(lambda x: list(x))\n",
    "#Obimac.groupby('Name',as_index=False)[['seqid']].aggregate(lambda x: list(x))\n",
    "# this groups by Name, but it it necessary for what I want? I just need a transcript \\t gene, where transcript \n",
    "# is the trinity transcript name, and 'gene' is the 'Name' value."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "dammit",
   "language": "python",
   "name": "dammit"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
