{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Select best dammit annotations for a trinity transcriptome. Create tx2gene maps for use with tximport; aggregate contigs with identical \"best\" annotations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "# requires dammit env (mine-->source activate dammit), or dammit installed in main env\n",
    "# if using an env: if you haven't done so yet, first install an ipy kernel in dammit env:\n",
    "# ipython kernel install --user --name dammit\n",
    "# then start new notebook using \"dammit\" kernel, or for existing nb, select \"switch kernel\" in the Kernel menu\n",
    "# voila! :D\n",
    "from dammit.fileio.gff3 import GFF3Parser"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Link Dammit and Trinity names\n",
    "Use Dammit Namemap to create Dammit_Name : Trinity Transcript/Gene map"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 315,
   "metadata": {},
   "outputs": [],
   "source": [
    "namemap = \"Trinity.fasta.dammit.namemap.csv\"\n",
    "trin2dammit = pd.read_csv(namemap)\n",
    "\n",
    "# create gene: trans map while we're at it\n",
    "trin2dammit['Trinity_transcript'] = trin2dammit['original'].str.split(' ', 1).str[0]\n",
    "trin2dammit['Trinity_gene'] = trin2dammit['Trinity_transcript'].str.rsplit('_i', 1).str[0] \n",
    "\n",
    "# drop 'original' trinity name; rename dammit name column\n",
    "trin2dammit.rename(index=str, columns={\"renamed\": \"Dammit_transcript\"}, inplace=True)\n",
    "trin2dammit.drop(columns=['original'], inplace=True)\n",
    "\n",
    "#print(trin2dammit.shape)\n",
    "#trin2dammit.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Parse Dammit GFF3\n",
    "Use Dammit GFF3 to read in annotations, and link with Trinity Names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 133,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/tessa/anaconda3/envs/dammit/lib/python3.6/site-packages/dammit/fileio/gff3.py:73: ParserWarning: Both a converter and dtype were specified for column attributes - only the converter will be used\n",
      "  dtype=dict(self.columns)):\n"
     ]
    }
   ],
   "source": [
    "annots_gff3_file = \"Trinity.fasta.dammit.gff3\"\n",
    "annots = GFF3Parser(filename=annots_gff3_file).read() # read in annotation gff3\n",
    "#print(annots.shape)\n",
    "#annots.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 134,
   "metadata": {},
   "outputs": [],
   "source": [
    "# merge annotation gff3 with dammit-trinity namemap\n",
    "annotsTrin = pd.merge(trin2dammit, annots, how='outer', left_on=\"Dammit_transcript\", right_on=\"seqid\")\n",
    "#print(annotsTrin.shape)\n",
    "#annotsTrin.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 132,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Of 150663 total transcripts:\n",
      "    63270 transcripts have at least one annotation\n",
      "    87393 transcripts have no annotations\n"
     ]
    }
   ],
   "source": [
    "#count transcripts/annotations & do a sanity check on the merge:\n",
    "numTranscripts = trin2dammit['Dammit_transcript'].nunique()\n",
    "numWithAnnots = annots['seqid'].nunique()\n",
    "numNoAnnots = numTranscripts - numWithAnnots\n",
    "numMergedTranscripts = annotsTrin['Dammit_transcript'].nunique()\n",
    "\n",
    "if (numMergedTranscripts != numTranscripts):\n",
    "    print('something went wrong during merge')\n",
    "else:\n",
    "    print('Of ' + str(numTranscripts) + ' total transcripts:')\n",
    "    print('    ' + str(numWithAnnots) + ' transcripts have at least one annotation')\n",
    "    print('    ' + str(numNoAnnots)  + ' transcripts have no annotations')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Select annotations with best e-values\n",
    "\n",
    "Create annotation csv with the 1. best e-val hit, and 2. best Obimac hit for each transcript"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 153,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Dammit_transcript</th>\n",
       "      <th>Trinity_transcript</th>\n",
       "      <th>Trinity_gene</th>\n",
       "      <th>Dbxref</th>\n",
       "      <th>ID</th>\n",
       "      <th>Name</th>\n",
       "      <th>Note</th>\n",
       "      <th>Parent</th>\n",
       "      <th>Target</th>\n",
       "      <th>accuracy</th>\n",
       "      <th>...</th>\n",
       "      <th>end</th>\n",
       "      <th>env_coords</th>\n",
       "      <th>phase</th>\n",
       "      <th>score</th>\n",
       "      <th>seqid</th>\n",
       "      <th>source</th>\n",
       "      <th>start</th>\n",
       "      <th>strand</th>\n",
       "      <th>trunc</th>\n",
       "      <th>type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Transcript_0</td>\n",
       "      <td>TRINITY_DN49110_c0_g1_i1</td>\n",
       "      <td>TRINITY_DN49110_c0_g1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Transcript_1</td>\n",
       "      <td>TRINITY_DN49156_c0_g1_i1</td>\n",
       "      <td>TRINITY_DN49156_c0_g1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Transcript_2</td>\n",
       "      <td>TRINITY_DN49121_c0_g1_i1</td>\n",
       "      <td>TRINITY_DN49121_c0_g1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Transcript_3</td>\n",
       "      <td>TRINITY_DN49116_c0_g1_i1</td>\n",
       "      <td>TRINITY_DN49116_c0_g1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>cds.Transcript_3.p2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Transcript_3.p2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>482.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Transcript_3</td>\n",
       "      <td>transdecoder</td>\n",
       "      <td>173.0</td>\n",
       "      <td>+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CDS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Transcript_3</td>\n",
       "      <td>TRINITY_DN49116_c0_g1_i1</td>\n",
       "      <td>TRINITY_DN49116_c0_g1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Transcript_3.p2.exon1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Transcript_3.p2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>482.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Transcript_3</td>\n",
       "      <td>transdecoder</td>\n",
       "      <td>0.0</td>\n",
       "      <td>+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>exon</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  Dammit_transcript        Trinity_transcript           Trinity_gene Dbxref  \\\n",
       "0      Transcript_0  TRINITY_DN49110_c0_g1_i1  TRINITY_DN49110_c0_g1    NaN   \n",
       "1      Transcript_1  TRINITY_DN49156_c0_g1_i1  TRINITY_DN49156_c0_g1    NaN   \n",
       "2      Transcript_2  TRINITY_DN49121_c0_g1_i1  TRINITY_DN49121_c0_g1    NaN   \n",
       "3      Transcript_3  TRINITY_DN49116_c0_g1_i1  TRINITY_DN49116_c0_g1    NaN   \n",
       "4      Transcript_3  TRINITY_DN49116_c0_g1_i1  TRINITY_DN49116_c0_g1    NaN   \n",
       "\n",
       "                      ID Name Note           Parent Target accuracy  ...   \\\n",
       "0                    NaN  NaN  NaN              NaN    NaN      NaN  ...    \n",
       "1                    NaN  NaN  NaN              NaN    NaN      NaN  ...    \n",
       "2                    NaN  NaN  NaN              NaN    NaN      NaN  ...    \n",
       "3    cds.Transcript_3.p2  NaN  NaN  Transcript_3.p2    NaN      NaN  ...    \n",
       "4  Transcript_3.p2.exon1  NaN  NaN  Transcript_3.p2    NaN      NaN  ...    \n",
       "\n",
       "     end env_coords  phase score         seqid        source  start strand  \\\n",
       "0    NaN        NaN    NaN   NaN           NaN           NaN    NaN    NaN   \n",
       "1    NaN        NaN    NaN   NaN           NaN           NaN    NaN    NaN   \n",
       "2    NaN        NaN    NaN   NaN           NaN           NaN    NaN    NaN   \n",
       "3  482.0        NaN    0.0   NaN  Transcript_3  transdecoder  173.0      +   \n",
       "4  482.0        NaN    NaN   NaN  Transcript_3  transdecoder    0.0      +   \n",
       "\n",
       "   trunc  type  \n",
       "0    NaN   NaN  \n",
       "1    NaN   NaN  \n",
       "2    NaN   NaN  \n",
       "3    NaN   CDS  \n",
       "4    NaN  exon  \n",
       "\n",
       "[5 rows x 22 columns]"
      ]
     },
     "execution_count": 153,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "annotsTrin.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 404,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(44964, 22)\n",
      "(46472, 26)\n",
      "(150663, 26)\n"
     ]
    }
   ],
   "source": [
    "# find best Obimac hit per transcript (using best evalue)\n",
    "Obimac = annotsTrin[annotsTrin['database'] == \"Obimac_refseq_protein.fa.gz\"]\n",
    "Obimac = Obimac.sort_values(by=['Trinity_transcript', 'score'], ascending=True).query('score < 1e-05').drop_duplicates(subset='Trinity_transcript')[['Trinity_transcript','Name', 'score', 'start', 'end']]\n",
    "#Obimac = Obimac.dropna(axis=0,how=\"all\")\n",
    "Obimac.rename(index=str, columns={'Name': 'Obimac_Name', 'score': 'Obimac_score', 'start': 'Obimac_start', 'end':'Obimac_end'}, inplace=True)\n",
    "\n",
    "Obimac.tail()\n",
    "\n",
    "#print(Obimac.shape)\n",
    "#Obimac.head()\n",
    "#list(Obimac['Name']) # if you want to read the names\n",
    "# find best Eval that is *not* an Obimac hit\n",
    "\n",
    "noObimac = annotsTrin[annotsTrin['database'] != 'Obimac_refseq_protein.fa.gz']\n",
    "bestEval = noObimac.sort_values(by=['Trinity_transcript', 'score'], ascending=True).query('score < 1e-05').drop_duplicates(subset='Trinity_transcript') #[['Dammit_transcript','Name', 'score', 'start', 'end','database']]\n",
    "#bestEval = bestEval.dropna(axis=0,how=\"all\")\n",
    "\n",
    "print(bestEval.shape)\n",
    "#bestEval.head()\n",
    "# merge bestEval and Obimac annotation info\n",
    "bestAnnots = pd.merge(bestEval, Obimac, how='outer', left_on=\"Trinity_transcript\", right_on=\"Trinity_transcript\")\n",
    "print(bestAnnots.shape)\n",
    "\n",
    "# to keep the unannotated transcripts, merge those back in\n",
    "bestAnnotFull = pd.merge(bestAnnots, trin2dammit[['Trinity_transcript']], how='outer', left_on=[\"Trinity_transcript\"],right_on=[\"Trinity_transcript\"])\n",
    "print(bestAnnotFull.shape)\n",
    "bestAnnotFull.to_csv(\"dammit_bestEvalperTranscript.csv\", index=False, sep='\\t')\n",
    "bestEval.to_csv(\"bestEvalpertranscript.csv\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The above keeps all identified trinity transcripts, and just reports the best annotation (determined by eval), as well as the best Obimac annotation (also det. by eval).\n",
    "\n",
    "However, there are undoubtedly some duplicate annotations, as there are v. likely not that many genes in D. opalescens. For differential expression, we're primarily interested in functional changes, and spreading the expression of a gene across different contigs with the exact same best annotation is not desirable. Can we collapse by best annotation?"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Create Tx2Gene files\n",
    "Create Transcript to Gene (tx2gene) files that assign transcripts to their annotated gene names,\n",
    "rather than just the Trinity gene name. If we want to keep unannotated transcripts, we can retain \n",
    "Trinity gene name for unannotated transcripts. These can be used with tximport to aggregate transcript-level\n",
    "quantification to gene-level counts for differential expression analysis."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 405,
   "metadata": {},
   "outputs": [],
   "source": [
    "bestAnnot_tx2gene = bestAnnots[['Trinity_transcript', 'Name']].dropna()\n",
    "#bestAnnots['Name'].nunique() # just checking that there are duplicate 'Name' column values. There are.\n",
    "bestAnnot_tx2gene.head()\n",
    "bestAnnot_tx2gene.to_csv('bestEval_tx2gene.txt', index=False, sep = '\\t')\n",
    "# do this using the Obimac best annotations instead:\n",
    "bestObimac_tx2gene = bestAnnots[['Trinity_transcript', 'Obimac_Name']].dropna()\n",
    "bestObimac_tx2gene.to_csv('bestObimac_tx2gene.txt', index=False, sep = '\\t') "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let's make a version that includes unannotated transcripts. Use Trinity gene info as gene name"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 406,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(150663, 26)\n",
      "# annotations using only best Evalue: 44964\n",
      "# annotations using best Evalue AND best Obimac Evalue: 46472\n"
     ]
    }
   ],
   "source": [
    "#remake df: (easier for testing)\n",
    "bestAnnotFull = pd.merge(bestAnnots, trin2dammit[['Trinity_transcript']], how='outer', left_on=[\"Trinity_transcript\"],right_on=[\"Trinity_transcript\"])\n",
    "print(bestAnnotFull.shape)\n",
    "# first, fill NA in the 'Name' column with Obimac annotations\n",
    "numAnnots = len(list(bestAnnotFull.Name.dropna()))\n",
    "print('# annotations using only best Evalue: ' + str(numAnnots))\n",
    "bestAnnotFull.Name = bestAnnotFull.Name.fillna(value=bestAnnotFull.Obimac_Name)\n",
    "numAnnots_wOb = len(list(bestAnnotFull.Name.dropna()))\n",
    "print('# annotations using best Evalue AND best Obimac Evalue: ' + str(numAnnots_wOb))\n",
    "bestEandObimac_tx2gene = bestAnnotFull[['Trinity_transcript', 'Name']].dropna()\n",
    "bestEandObimac_tx2gene.to_csv('bestEval_Obimac_tx2gene.txt', index=False, sep = '\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 409,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(150663, 2)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Dammit_transcript</th>\n",
       "      <th>Trinity_transcript</th>\n",
       "      <th>Trinity_gene</th>\n",
       "      <th>Dbxref</th>\n",
       "      <th>ID</th>\n",
       "      <th>Name</th>\n",
       "      <th>Note</th>\n",
       "      <th>Parent</th>\n",
       "      <th>Target</th>\n",
       "      <th>accuracy</th>\n",
       "      <th>...</th>\n",
       "      <th>seqid</th>\n",
       "      <th>source</th>\n",
       "      <th>start</th>\n",
       "      <th>strand</th>\n",
       "      <th>trunc</th>\n",
       "      <th>type</th>\n",
       "      <th>Obimac_Name</th>\n",
       "      <th>Obimac_score</th>\n",
       "      <th>Obimac_start</th>\n",
       "      <th>Obimac_end</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Transcript_55709</td>\n",
       "      <td>TRINITY_DN10002_c0_g1_i1</td>\n",
       "      <td>TRINITY_DN10002_c0_g1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>homology:127783</td>\n",
       "      <td>LotgiP220800</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>LotgiP220800 1222 1355 +</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>Transcript_55709</td>\n",
       "      <td>LAST</td>\n",
       "      <td>1.0</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>translated_nucleotide_match</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Transcript_55702</td>\n",
       "      <td>TRINITY_DN10017_c0_g1_i1</td>\n",
       "      <td>TRINITY_DN10017_c0_g1</td>\n",
       "      <td>\"Pfam:PF00085.16\"</td>\n",
       "      <td>homology:32447</td>\n",
       "      <td>Thioredoxin</td>\n",
       "      <td>Thioredoxin</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Thioredoxin 5 74 +</td>\n",
       "      <td>0.92</td>\n",
       "      <td>...</td>\n",
       "      <td>Transcript_55702</td>\n",
       "      <td>HMMER</td>\n",
       "      <td>141.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>protein_hmm_match</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Transcript_55699</td>\n",
       "      <td>TRINITY_DN10022_c0_g1_i1</td>\n",
       "      <td>TRINITY_DN10022_c0_g1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>homology:127777</td>\n",
       "      <td>SMAR010045-PA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SMAR010045-PA 58 105 +</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>Transcript_55699</td>\n",
       "      <td>LAST</td>\n",
       "      <td>403.0</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>translated_nucleotide_match</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Transcript_55695</td>\n",
       "      <td>TRINITY_DN10025_c0_g1_i1</td>\n",
       "      <td>TRINITY_DN10025_c0_g1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>homology:127774</td>\n",
       "      <td>OABI005503-RA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>OABI005503-RA 401 466 +</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>Transcript_55695</td>\n",
       "      <td>LAST</td>\n",
       "      <td>514.0</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>translated_nucleotide_match</td>\n",
       "      <td>gi|961083531|ref|XP_014769378.1| PREDICTED: BR...</td>\n",
       "      <td>3.400000e-55</td>\n",
       "      <td>27.0</td>\n",
       "      <td>236.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Transcript_55697</td>\n",
       "      <td>TRINITY_DN10032_c0_g1_i1</td>\n",
       "      <td>TRINITY_DN10032_c0_g1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>homology:127776</td>\n",
       "      <td>CapteP182594</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CapteP182594 1098 1233 +</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>Transcript_55697</td>\n",
       "      <td>LAST</td>\n",
       "      <td>14.0</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>translated_nucleotide_match</td>\n",
       "      <td>gi|961119188|ref|XP_014781878.1| PREDICTED: ca...</td>\n",
       "      <td>4.200000e-70</td>\n",
       "      <td>0.0</td>\n",
       "      <td>129.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 26 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  Dammit_transcript        Trinity_transcript           Trinity_gene  \\\n",
       "0  Transcript_55709  TRINITY_DN10002_c0_g1_i1  TRINITY_DN10002_c0_g1   \n",
       "1  Transcript_55702  TRINITY_DN10017_c0_g1_i1  TRINITY_DN10017_c0_g1   \n",
       "2  Transcript_55699  TRINITY_DN10022_c0_g1_i1  TRINITY_DN10022_c0_g1   \n",
       "3  Transcript_55695  TRINITY_DN10025_c0_g1_i1  TRINITY_DN10025_c0_g1   \n",
       "4  Transcript_55697  TRINITY_DN10032_c0_g1_i1  TRINITY_DN10032_c0_g1   \n",
       "\n",
       "              Dbxref               ID           Name         Note Parent  \\\n",
       "0                NaN  homology:127783   LotgiP220800          NaN    NaN   \n",
       "1  \"Pfam:PF00085.16\"   homology:32447    Thioredoxin  Thioredoxin    NaN   \n",
       "2                NaN  homology:127777  SMAR010045-PA          NaN    NaN   \n",
       "3                NaN  homology:127774  OABI005503-RA          NaN    NaN   \n",
       "4                NaN  homology:127776   CapteP182594          NaN    NaN   \n",
       "\n",
       "                     Target accuracy    ...                 seqid source  \\\n",
       "0  LotgiP220800 1222 1355 +      NaN    ...      Transcript_55709   LAST   \n",
       "1        Thioredoxin 5 74 +     0.92    ...      Transcript_55702  HMMER   \n",
       "2    SMAR010045-PA 58 105 +      NaN    ...      Transcript_55699   LAST   \n",
       "3   OABI005503-RA 401 466 +      NaN    ...      Transcript_55695   LAST   \n",
       "4  CapteP182594 1098 1233 +      NaN    ...      Transcript_55697   LAST   \n",
       "\n",
       "   start strand  trunc                         type  \\\n",
       "0    1.0      -    NaN  translated_nucleotide_match   \n",
       "1  141.0    NaN    NaN            protein_hmm_match   \n",
       "2  403.0      -    NaN  translated_nucleotide_match   \n",
       "3  514.0      -    NaN  translated_nucleotide_match   \n",
       "4   14.0      -    NaN  translated_nucleotide_match   \n",
       "\n",
       "                                         Obimac_Name  Obimac_score  \\\n",
       "0                                                NaN           NaN   \n",
       "1                                                NaN           NaN   \n",
       "2                                                NaN           NaN   \n",
       "3  gi|961083531|ref|XP_014769378.1| PREDICTED: BR...  3.400000e-55   \n",
       "4  gi|961119188|ref|XP_014781878.1| PREDICTED: ca...  4.200000e-70   \n",
       "\n",
       "   Obimac_start Obimac_end  \n",
       "0           NaN        NaN  \n",
       "1           NaN        NaN  \n",
       "2           NaN        NaN  \n",
       "3          27.0      236.0  \n",
       "4           0.0      129.0  \n",
       "\n",
       "[5 rows x 26 columns]"
      ]
     },
     "execution_count": 409,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Then, fill remaining NA values in the 'Name' column with the Trinity Gene name:\n",
    "bestAnnotFull.Name = bestAnnotFull.Name.fillna(value=bestAnnotFull.Trinity_gene)\n",
    "bestAnnotFull_tx2gene = bestAnnotFull[['Trinity_transcript', 'Name']]\n",
    "print(bestAnnotFull_tx2gene.shape)\n",
    "bestAnnotFull_tx2gene.to_csv('bestEvalFull_tx2gene.txt', index=False, sep = '\\t')\n",
    "bestAnnotFull.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Not used: Aggregate by identical best eval hits:\n",
    "# Collapse using best Eval hit:\n",
    "#bestAnnots.groupby('Name',as_index=False)[['Dammit_transcript']].aggregate(lambda x: list(x))\n",
    "#Obimac.groupby('Name',as_index=False)[['seqid']].aggregate(lambda x: list(x))\n",
    "# this groups by Name, but it it necessary for what I want? I just need a transcript \\t gene, where transcript \n",
    "# is the trinity transcript name, and 'gene' is the 'Name' value."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "dammit",
   "language": "python",
   "name": "dammit"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
